File: benchmarks/pandas/bench_errors_extended.py
"""
Benchmark: pd.errors namespace (extended) — remaining pandas-compatible
error/warning classes not covered by bench_errors.py.

Mirrors tsb's `errors` namespace additions: AbstractMethodError,
AccessorRegistrationWarning, AttributeConflictWarning, CSSWarning,
ChainedAssignmentError, DatabaseError, DtypeWarning, DuplicateLabelError,
InvalidColumnName, InvalidComparison, InvalidUseOfBooleanIndex,
InvalidVersion, LossySetitemError, NullFrequencyError, NumbaUtilError,
OutOfBoundsDatetime, OutOfBoundsTimedelta, ParserWarning,
PossibleDataLossError, PossiblePrecisionLoss, SpecificationError,
UnsupportedFunctionCall, ValueLabelTypeMismatch.

A handful of these names (InvalidColumnName, InvalidUseOfBooleanIndex,
LossySetitemError, ValueLabelTypeMismatch) are tsb-specific extensions with
no direct `pandas.errors` counterpart across all pandas versions, so a local
fallback class is used for those to keep the benchmark portable.

Outputs JSON: {"function": "errors_extended", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""

import json
import time

import pandas.errors as pd_errors


class _FallbackError(ValueError):
    """Local stand-in for tsb-only error classes with no pandas.errors match."""


InvalidColumnName = getattr(pd_errors, "InvalidColumnName", _FallbackError)
InvalidUseOfBooleanIndex = getattr(pd_errors, "InvalidUseOfBooleanIndex", _FallbackError)
LossySetitemError = getattr(pd_errors, "LossySetitemError", _FallbackError)
ValueLabelTypeMismatch = getattr(pd_errors, "ValueLabelTypeMismatch", _FallbackError)

WARMUP = 5
ITERATIONS = 200


def _run():
    e1 = pd_errors.AbstractMethodError(object())
    e2 = pd_errors.AccessorRegistrationWarning("shadowed accessor")
    e3 = pd_errors.AttributeConflictWarning("conflicting attribute")
    e4 = pd_errors.CSSWarning("bad css rule")
    e5 = pd_errors.ChainedAssignmentError()
    e6 = pd_errors.DatabaseError("connection failed")
    e7 = pd_errors.DtypeWarning("mismatched dtypes")
    e8 = pd_errors.DuplicateLabelError("Index has duplicates")
    e9 = InvalidColumnName("bad column")
    e10 = pd_errors.InvalidComparison("incompatible types")
    e11 = InvalidUseOfBooleanIndex("bad boolean index")
    e12 = pd_errors.InvalidVersion("not.a.version")
    e13 = LossySetitemError("would lose precision")
    e14 = pd_errors.NullFrequencyError("null frequency")
    e15 = pd_errors.NumbaUtilError("numba failure")
    e16 = pd_errors.OutOfBoundsDatetime("out of bounds timestamp")
    e17 = pd_errors.OutOfBoundsTimedelta("out of bounds timedelta")
    e18 = pd_errors.ParserWarning("falling back to python engine")
    e19 = pd_errors.PossibleDataLossError("mode='w' will overwrite")
    e20 = pd_errors.PossiblePrecisionLoss("float64 -> float32")
    e21 = pd_errors.SpecificationError("nested renamer is not supported")
    e22 = pd_errors.UnsupportedFunctionCall("numpy operation not supported")
    e23 = ValueLabelTypeMismatch("mismatched value/label types")

    _a = isinstance(e1, Exception)
    _b = isinstance(e2, pd_errors.AccessorRegistrationWarning)
    _c = type(e3).__name__ == "AttributeConflictWarning"
    _d = type(e4).__name__ == "CSSWarning"
    _e = "copy" in str(e5) or True
    _f = isinstance(e6, pd_errors.DatabaseError)
    _g = type(e7).__name__ == "DtypeWarning"
    _h = isinstance(e8, ValueError)
    _i = type(e9).__name__ in ("InvalidColumnName", "_FallbackError")
    _j = isinstance(e10, TypeError)
    _k = isinstance(e11, (IndexError, ValueError))
    _l = isinstance(e12, ValueError)
    _m = type(e13).__name__ in ("LossySetitemError", "_FallbackError")
    _n = isinstance(e14, ValueError)
    _o = type(e15).__name__ == "NumbaUtilError"
    _p = isinstance(e16, ValueError)
    _q = isinstance(e17, ValueError)
    _r = type(e18).__name__ == "ParserWarning"
    _s = type(e19).__name__ == "PossibleDataLossError"
    _t = type(e20).__name__ == "PossiblePrecisionLoss"
    _u = isinstance(e21, ValueError)
    _v = isinstance(e22, ValueError)
    _w = type(e23).__name__ in ("ValueLabelTypeMismatch", "_FallbackError")
    return [
        _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n, _o, _p, _q, _r, _s, _t, _u, _v, _w,
    ]


for _ in range(WARMUP):
    _run()

start = time.perf_counter()
for _ in range(ITERATIONS):
    _run()
total_ms = (time.perf_counter() - start) * 1000

print(
    json.dumps(
        {
            "function": "errors_extended",
            "mean_ms": total_ms / ITERATIONS,
            "iterations": ITERATIONS,
            "total_ms": total_ms,
        }
    )
)

File: benchmarks/tsb/bench_errors_extended.ts
/**
 * Benchmark: pd.errors namespace (extended) — remaining pandas-compatible
 * error/warning classes not covered by bench_errors.ts.
 *
 * Covers 23 additional classes from tsb's `errors` namespace:
 *   - errors.AbstractMethodError, errors.AccessorRegistrationWarning
 *   - errors.AttributeConflictWarning, errors.CSSWarning
 *   - errors.ChainedAssignmentError, errors.DatabaseError
 *   - errors.DtypeWarning, errors.DuplicateLabelError
 *   - errors.InvalidColumnName, errors.InvalidComparison
 *   - errors.InvalidUseOfBooleanIndex, errors.InvalidVersion
 *   - errors.LossySetitemError, errors.NullFrequencyError
 *   - errors.NumbaUtilError, errors.OutOfBoundsDatetime
 *   - errors.OutOfBoundsTimedelta, errors.ParserWarning
 *   - errors.PossibleDataLossError, errors.PossiblePrecisionLoss
 *   - errors.SpecificationError, errors.UnsupportedFunctionCall
 *   - errors.ValueLabelTypeMismatch
 *   - instanceof checks and .name/.message property access
 *
 * Outputs JSON: {"function": "errors_extended", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { errors } from "../../src/index.ts";

const WARMUP = 5;
const ITERATIONS = 200;

function run(): void {
  const e1 = new errors.AbstractMethodError("MyClass");
  const e2 = new errors.AccessorRegistrationWarning("shadowed accessor");
  const e3 = new errors.AttributeConflictWarning("conflicting attribute");
  const e4 = new errors.CSSWarning("bad css rule");
  const e5 = new errors.ChainedAssignmentError();
  const e6 = new errors.DatabaseError("connection failed");
  const e7 = new errors.DtypeWarning("mismatched dtypes");
  const e8 = new errors.DuplicateLabelError();
  const e9 = new errors.InvalidColumnName("bad column");
  const e10 = new errors.InvalidComparison("incompatible types");
  const e11 = new errors.InvalidUseOfBooleanIndex("bad boolean index");
  const e12 = new errors.InvalidVersion("not.a.version");
  const e13 = new errors.LossySetitemError("would lose precision");
  const e14 = new errors.NullFrequencyError();
  const e15 = new errors.NumbaUtilError("numba failure");
  const e16 = new errors.OutOfBoundsDatetime();
  const e17 = new errors.OutOfBoundsTimedelta();
  const e18 = new errors.ParserWarning("falling back to python engine");
  const e19 = new errors.PossibleDataLossError("mode='w' will overwrite");
  const e20 = new errors.PossiblePrecisionLoss("float64 -> float32");
  const e21 = new errors.SpecificationError("nested renamer is not supported");
  const e22 = new errors.UnsupportedFunctionCall("numpy operation not supported");
  const e23 = new errors.ValueLabelTypeMismatch("mismatched value/label types");

  const _a = e1 instanceof Error;
  const _b = e2 instanceof errors.AccessorRegistrationWarning;
  const _c = e3.name === "AttributeConflictWarning";
  const _d = e4.name === "CSSWarning";
  const _e = e5.message.includes("copy");
  const _f = e6 instanceof errors.DatabaseError;
  const _g = e7.name === "DtypeWarning";
  const _h = e8 instanceof errors.ValueError;
  const _i = e9.name === "InvalidColumnName";
  const _j = e10 instanceof TypeError;
  const _k = e11 instanceof errors.IndexError;
  const _l = e12 instanceof errors.ValueError;
  const _m = e13.name === "LossySetitemError";
  const _n = e14 instanceof errors.ValueError;
  const _o = e15.name === "NumbaUtilError";
  const _p = e16 instanceof errors.ValueError;
  const _q = e17 instanceof errors.ValueError;
  const _r = e18.name === "ParserWarning";
  const _s = e19.name === "PossibleDataLossError";
  const _t = e20.name === "PossiblePrecisionLoss";
  const _u = e21 instanceof errors.ValueError;
  const _v = e22 instanceof errors.ValueError;
  const _w = e23.name === "ValueLabelTypeMismatch";
  void [
    _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n, _o, _p, _q, _r, _s, _t, _u, _v, _w,
  ];
}

for (let i = 0; i < WARMUP; i++) run();

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) run();
const total_ms = performance.now() - start;

console.log(
  JSON.stringify({
    function: "errors_extended",
    mean_ms: total_ms / ITERATIONS,
    iterations: ITERATIONS,
    total_ms,
  }),
);
